File: android/core/domain/src/main/java/com/bobaplaybook/core/domain/model/Card.kt
package com.bobaplaybook.core.domain.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * BOBA card catalog model. Mirrors `BOBAPlaybook/Models/Card.swift`
 * field-for-field; deviations are CALLED OUT in comments.
 *
 * The catalog is bundled in `app/src/main/assets/data/cards.json` via
 * the Gradle copy task in [ANDROID-DEV.md §13.3]. ~17,974 entries
 * decode at app start in a two-phase pipeline (DECISIONS.md #014).
 *
 * `ignoreUnknownKeys = true` is set on the decoder so the catalog can
 * grow new fields without breaking the Android client (mirrors iOS
 * Codable's "extra keys are silently dropped" semantics).
 *
 * `bobaId` is computed at runtime — same 4-field formula as `scripts/
 * boba_id.py` and `Card.swift::bobaId`. Sealed Products use `name` when
 * `hero` is empty. Trailing dashes are intentional and stable. Verified
 * 17,739 unique values across the bundle on iOS.
 *
 * Don't add an Android-only field here without first proposing it for
 * the canonical schema (`docs/CARD_SCHEMA.md`). Cross-platform drift is
 * the bug.
 */
@Serializable
data class Card(
    val bvId: Int? = null,
    val cardNumber: String,
    val name: String,
    val hero: String = "",
    val cardType: String,
    // Sealed Products (Booster Box, Blaster Box, Case, etc.) have
    // `element: null` in cards.json — they're not playable cards
    // with a weapon. Without the empty-string default + the
    // `coerceInputValues = true` Json config, kotlinx.serialization
    // would fail the FIRST null-element row and abort the full
    // catalog decode — leaving the in-memory catalog stuck at the
    // 506 cards from cards-head.json. That silently dropped EVERY
    // user_cards row in the Collection JOIN because almost all real
    // cards live past the head-bundle window. Default-to-"" keeps
    // every downstream `card.element.uppercase()` / `.lowercase()`
    // call site working without a sweep; sealed-product call sites
    // already guard via `card.isSealed`.
    val element: String = "",           // UPPERCASE in JSON when present, "" for Sealed Products, mixed-case in UI
    val set: String,
    @SerialName("subSet") val subSet: String? = null,
    val treatment: String? = null,
    val variation: String? = null,
    val release: String? = null,

    // Stats
    val power: Int? = null,
    val cost: Int? = null,
    val dbs: Int? = null,                // Damage Bonus / Score (Plays only)
    val hitPoints: Int? = null,
    val hd: Int? = null,                 // Heroic Damage

    // Catalog flags
    val rarityLabel: String? = null,
    val rarityTier: String? = null,
    val rookieInspired: Boolean = false,

    // Image fields — mirrors iOS Card.imageFile being mutable for runtime
    // override (admin upload → `setAppliedOverride` → in-place mutation).
    // Kotlin data classes don't have `var` properties as ergonomic as
    // Swift; the override path is handled in the CardRepository instead
    // of mutating the model. See ANDROID-DEV.md §11.4 for the swap
    // pattern.
    val imageFile: String? = null,
    val imageSource: String? = null,
    val imageAvailable: Boolean = false,
    /**
     * Frozen legacy field — populated for cards added to the catalog
     * before 2026-05-23. Used SOLELY as the destination of the per-card
     * "View on Radish" external-link button per Radish's email-stated
     * allowance for "ordinary user-facing linking." Null for new cards;
     * the button falls back to the Radish homepage. NEVER passed to any
     * Worker / matcher / pricing lookup — that automation is prohibited.
     */
    val radishUrl: String? = null,
    val bvUrl: String? = null,
    /**
     * Card-art-OCR'd print run, populated for 464 of 17,974 catalog
     * cards (PRICING_PLAYBOOK §6.4 Feature 0). `5`, `10`, `25`, or
     * `50` typically — observed serialization, the hardest scarcity
     * signal we have. Null for the un-numbered majority. DECISIONS.md
     * #061 makes this the ONLY source of [printRunLabel] — the prior
     * weapon→count switch (#028 assumption) shipped wrong values on
     * 174+ FIRE cards because FIRE + ICE each appear at BOTH /5 AND
     * /50 in the real OCR data.
     */
    val printRun: Int? = null,

    /** The real-world athlete this card's hero is inspired by. */
    val athleteInspiration: String? = null,

    /**
     * Per-card searchable aliases — e.g. `["Skeeball"]` on Skeee cards
     * so users typing the printed-on-card name (which differs from the
     * catalog hero) still find them. Merged into the haystack at
     * search time by CardSearch.
     */
    val searchAliases: List<String>? = null,

    // Play-only fields
    val persistent: List<PersistentEffectSpec> = emptyList(),
    val abilityText: String? = null,
    val bonusText: String? = null,
    @SerialName("bobaId") val bobaIdField: String? = null,     // when present in JSON, prefer this over computed

    // Play subtype flags. iOS Card.swift lines 31-32. Used by
    // CardFormatEligibility to surface "Bonus Play" / "HTD Play" badges
    // that are toggled OFF in certain event formats (Spec Playmaker,
    // Brawl Playmaker, Tecmo Bowl).
    @SerialName("isBonusPlay") val isBonusPlay: Boolean? = null,
    @SerialName("isHTD")       val isHTD: Boolean? = null,

    /**
     * Tick 189 — Discord backlog #7. Catalog has `isInspiredInk` on
     * every card; iOS Card.swift mirrors it as a non-optional Bool.
     * Inspired Ink = Serialized variant with weapon-tied print runs
     * (Hex /5, Glow /10, Fire /25, Ice /50 per DECISIONS.md #028).
     */
    @SerialName("isInspiredInk") val isInspiredInk: Boolean = false,
) {
    /**
     * Canonical card identifier — matches `scripts/boba_id.py` v3
     * formula and iOS `Card.bobaId`. CLAUDE.md "One ID per Card" — this
     * is the primary key.
     *
     * 5-field formula:
     *   cardNumber-(hero|name)-treatment-variation-element
     *
     * The 5th field is the card's WEAPON (catalog stores it under the
     * legacy field name `element` per DECISIONS.md #027). Added
     * 2026-05-25 to disambiguate FIRE-weapon vs GLOW-weapon variant
     * siblings that share otherwise-identical (cardNumber, hero,
     * treatment, variation).
     *
     * Sealed products fall back to `name` when `hero` is empty.
     * Trailing dashes are intentional and stable.
     */
    val bobaId: String
        get() = bobaIdField ?: run {
            val identifier = hero.ifEmpty { name }
            "$cardNumber-$identifier-${treatment.orEmpty()}-${variation.orEmpty()}-$element"
        }

    /** True for sealed-product entries (no hero, has a name like "Booster Box"). */
    val isSealed: Boolean
        get() = cardType.equals("Sealed Product", ignoreCase = true) || hero.isEmpty()

    /** Card-type classifiers (mirrors iOS Card.swift). */
    val isHero: Boolean   get() = cardType.equals("Hero", ignoreCase = true)
    val isPlay: Boolean   get() = cardType.equals("Play", ignoreCase = true)
    val isHotDog: Boolean get() = cardType.equals("HotDog", ignoreCase = true) ||
                                  cardType.contains("Hot Dog", ignoreCase = true)

    /** User-facing display name. Heroes show their hero name; sealed shows product name. */
    val displayName: String
        get() = if (hero.isNotEmpty()) hero else name

    /**
     * Print-run label for cards — reads the catalog's real [printRun]
     * field (populated for 464 cards via card-art OCR; PRICING_PLAYBOOK
     * §6.4 Feature 0). Returns null when there's no OCR'd value, which
     * is most cards. The prior weapon→count switch (DECISIONS.md #028
     * assumption — FIRE→/25, ICE→/50, etc.) is retired by #061: the
     * OCR data shows FIRE + ICE each appear at BOTH /5 AND /50, so
     * the weapon→count mapping shipped wrong values on 174+ FIRE
     * cards alone. Anything not on the card art itself is a guess,
     * and we ship the truth or nothing.
     */
    val printRunLabel: String?
        get() = printRun?.takeIf { it > 0 }?.let { "/$it" }
}

/**
 * Persistent-effect spec (Play cards). Mirrors the iOS `PersistentEffect`
 * struct that DECISIONS.md #030 documents.
 *
 * Lean placeholder for now — the Practice executor (M5.5) ports the
 * full state-machine engine and will expand this.
 */
@Serializable
data class PersistentEffectSpec(
    val trigger: String,
    val scope: String? = null,
    val n: Int? = null,
    val effect: String? = null,
)

File: pipeline/recognition/CardRecognitionCLI/Sources/CardRecognitionCLI/Mirror/Card.swift
import Foundation
import CoreTransferable
import UniformTypeIdentifiers

/// `nonisolated` overrides the project's default-MainActor isolation
/// so this Sendable value type can be read from background actors
/// (notably the grid-scan TaskGroup, which calls into ScanMatching
/// off MainActor for parallelism). The struct is Sendable; mutation
/// is by-value-copy so the lone `var` (imageFile — updated by
/// CardStore.applyRuntimeImageOverrides from card_image_overrides
/// applied rows) is safe across actors.
nonisolated struct Card: Codable, Identifiable, Hashable, Sendable {
    let bvId: Int?
    let cardNumber: String
    let name: String
    let hero: String          // "" for sealed products
    let cardType: String
    let set: String
    let subSet: String?
    let variation: String?
    let treatment: String?
    /// Canonical release label, normalized from `set` (drops the
    /// "Edition" suffix where applicable). Matches the names used in
    /// BoBA's published DBS update and the bobaleagues CSV format
    /// (Alpha / Alpha Update / Griffey / Alpha Blast / Promo / etc.).
    /// Empty string for legacy records that haven't been re-bundled.
    let release: String
    let element: String       // "NONE" for sealed products
    let power: Int?
    let playCost: Int?
    let isBonusPlay: Bool?
    let isHTD: Bool?                  // Play cards only — marks "Home Team Discount" HTD variants
    let playAbility: String?
    let dbs: Int?                     // Deck Balancing System score (Plays only, nil for Heroes/HotDogs/Sealed)
    let dbsTier: String?              // "Low" | "Medium" | "High" | "Very High" (Plays only)
    let athleteInspiration: String?
    /// Per-card searchable aliases — e.g. ["Skeeball"] on Skeee cards
    /// so users typing the printed-on-card name (which differs from
    /// the catalog hero name) still find the card. Word-split and
    /// merged into the haystack at match time by CardSearch.
    let searchAliases: [String]?
    let isInspiredInk: Bool
    /// True when the hero's inspiration athlete was in their rookie season
    /// at print time. Non-Hero rows decode as false.
    let rookieInspired: Bool
    var imageFile: String?
    let imageSource: String?
    let imageAvailable: Bool
    /// Frozen legacy field — populated for cards in the catalog before
    /// 2026-05-23. Used SOLELY as the destination of the per-card
    /// "View on Radish" external-link button per Radish's email-stated
    /// allowance for "ordinary user-facing linking." Nil for new
    /// cards; the button falls back to the Radish homepage. NEVER
    /// passed to any Worker / matcher / pricing lookup — that
    /// automation is prohibited.
    let radishUrl: String?

    // Sealed product fields (nil for regular cards)
    let sealedProductId: String?
    let productType: String?
    let packsPerBox: Int?
    let cardsPerPack: Int?
    let totalCards: Int?
    let msrp: Double?
    let upc: String?
    let highlights: [String]?
    let caseQuantity: Int?
    let ebaySearchQuery: String?

    var isSealed:  Bool { cardType == "Sealed Product" }
    var isHero:    Bool { cardType == "Hero" }
    var isPlay:    Bool { cardType == "Play" }
    var isHotDog:  Bool { cardType == "HotDog" }

    /// Display name: for non-Hero cards whose name/hero is non-ASCII (e.g. Japanese "怪獣焼き"),
    /// fall back to the variation field which holds the readable English name.
    var displayName: String {
        guard !isHero else { return name }
        let raw = name
        let hasNonASCII = raw.unicodeScalars.contains { $0.value > 127 }
        if hasNonASCII, let v = variation, v.unicodeScalars.allSatisfy({ $0.value <= 127 }) {
            return v
        }
        return raw
    }

    var playCostLabel: String? {
        guard let cost = playCost else { return nil }
        return cost == 0 ? "FREE" : "\(cost) HD"
    }

    // Rarity derived from treatment field
    var rarityTier: Int {
        guard let t = treatment?.lowercased() else { return 0 }
        if t.contains("kanji")        { return 5 }
        if t.contains("superfoil") || isInspiredInk { return 4 }
        if t.contains("blizzard")     { return 3 }
        if t.contains("battlefoil") || t.contains("logofoil") { return 2 }
        if t.contains("blast") || t.contains("paper") { return 1 }
        return 0
    }

    var rarityLabel: String {
        switch rarityTier {
        case 5: return "Kanjifoil"
        case 4: return isInspiredInk ? "Inspired Ink" : "Superfoil"
        case 3: return "Blizzard"
        case 2: return treatment?.lowercased().contains("logofoil") == true ? "Logofoil" : "Battlefoil"
        case 1: return treatment?.lowercased().contains("blast") == true ? "Blast" : "Paper"
        default: return "Base Set"
        }
    }

    /// Tick 197 — Discord backlog #7 (Android tick 189 parity).
    /// Optional print-run label for at-a-glance scarcity. Returns nil
    /// for the typical 99% card. Source of truth: DECISIONS.md #028
    /// "Inspired Ink = Serialized with weapon-tied print numbers:
    /// Hex /5, Glow /10, Fire /25, Ice /50."  Superfoil is the
    /// next-rarest non-numbered treatment.
    var printRunLabel: String? {
        if isInspiredInk {
            switch element.uppercased() {
            case "HEX":  return "/5"
            case "GLOW": return "/10"
            case "FIRE": return "/25"
            case "ICE":  return "/50"
            default:     return "Serial"   // Steel/Gum/Brawl/Super Inspired Ink — print run not public
            }
        }
        if treatment?.lowercased().contains("superfoil") == true { return "SSP" }
        return nil
    }

    // Stable unique id — v3 formula matches boba_id.py:
    // "{cardNumber}-{hero or name}-{treatment??''}-{variation??''}-{element}".
    // 5th field is the weapon (DECISIONS.md #057 / 2026-05-25).
    // For Sealed Products (no hero) the `name` field stands in so
    // the id matches the catalog's canonical bobaId field + the
    // `Card.bobaId` getter below. Without the name fallback,
    // `card.id` for sealed produced an "old form" string
    // ("SEALED-foo---Bundle") that DIDN'T match the canonical
    // bobaId stored in cards.json — which is what the catalog
    // index `cardsById` is keyed on. User_cards rows written
    // with the canonical bobaId then failed to resolve, and
    // sealed products rendered as their raw bobaId in Collection
    // (Ben 2026-05-22).
    /// Hashable / Identifiable key. Must match the stored `bobaId`
    /// field (v3 5-field formula per DECISIONS.md #057) so the iOS
    /// catalog index `cardsById` agrees with Supabase user_cards rows
    /// + shared-URL targets. The v2 4-field formula collided on
    /// FIRE/GLOW weapon-variant pairs (101 collisions across GLBF +
    /// RAD), which crashed `Dictionary(uniqueKeysWithValues:)` in
    /// DeckBuilderStore.loadTemplate and elsewhere.
    ///
    /// `nonisolated` because the project default isolation is MainActor
    /// (per memory feedback_project_default_mainactor_isolation) but
    /// pure value-derived computed properties are safe across actors
    /// and Hashable conformance in Sets / Dictionary keys requires it.
    nonisolated var id: String { bobaId }

    nonisolated func hash(into hasher: inout Hasher) { hasher.combine(id) }
    nonisolated static func == (lhs: Card, rhs: Card) -> Bool { lhs.id == rhs.id }

    // Custom decoder handles nullable hero/element/imageAvailable for sealed products
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        bvId               = try c.decodeIfPresent(Int.self,      forKey: .bvId)
        cardNumber         = try c.decode(String.self,             forKey: .cardNumber)
        name               = try c.decode(String.self,             forKey: .name)
        hero               = try c.decodeIfPresent(String.self,    forKey: .hero)      ?? ""
        cardType           = try c.decode(String.self,             forKey: .cardType)
        set                = try c.decode(String.self,             forKey: .set)
        subSet             = try c.decodeIfPresent(String.self,    forKey: .subSet)
        variation          = try c.decodeIfPresent(String.self,    forKey: .variation)
        treatment          = try c.decodeIfPresent(String.self,    forKey: .treatment)
        release            = try c.decodeIfPresent(String.self,    forKey: .release) ?? ""
        // Default to empty string (NOT "NONE") for null element so the
        // v3 bobaId formula matches the canonical Python/Android source:
        // Python uses `card.get("element") or ""` → empty; Android uses
        // `val element: String = ""`. Sealed products in the catalog
        // store empty element; iOS defaulting to "NONE" caused the
        // computed bobaId to end in "-NONE" while stored ended in "-",
        // producing a Card.id mismatch that hid sealed-product art +
        // metadata in Collection / Find.
        element            = try c.decodeIfPresent(String.self,    forKey: .element)   ?? ""
        power              = try c.decodeIfPresent(Int.self,       forKey: .power)
        // playCost is an Int for Play cards (0–6 Hot Dogs) but sealed products
        // incorrectly store their MSRP price here as a Double. Decode flexibly.
        if let intVal = try? c.decodeIfPresent(Int.self, forKey: .playCost) {
            playCost = intVal
        } else if let dblVal = try? c.decodeIfPresent(Double.self, forKey: .playCost) {
            playCost = Int(dblVal)
        } else {
            playCost = nil
        }
        isBonusPlay        = try c.decodeIfPresent(Bool.self,       forKey: .isBonusPlay)
        isHTD              = try c.decodeIfPresent(Bool.self,       forKey: .isHTD)
        playAbility        = try c.decodeIfPresent(String.self,    forKey: .playAbility)
        dbs                = try c.decodeIfPresent(Int.self,       forKey: .dbs)
        dbsTier            = try c.decodeIfPresent(String.self,    forKey: .dbsTier)
        athleteInspiration = try c.decodeIfPresent(String.self,    forKey: .athleteInspiration)
        searchAliases      = try c.decodeIfPresent([String].self,  forKey: .searchAliases)
        isInspiredInk      = try c.decodeIfPresent(Bool.self,      forKey: .isInspiredInk) ?? false
        rookieInspired     = try c.decodeIfPresent(Bool.self,      forKey: .rookieInspired) ?? false
        let file           = try c.decodeIfPresent(String.self,    forKey: .imageFile)
        imageFile          = file
        imageSource        = try c.decodeIfPresent(String.self,    forKey: .imageSource)
        // imageAvailable may be null (sealed); derive from imageFile presence as fallback
        imageAvailable     = try c.decodeIfPresent(Bool.self,      forKey: .imageAvailable)
                             ?? !(file?.isEmpty ?? true)
        radishUrl          = try c.decodeIfPresent(String.self,    forKey: .radishUrl)
        sealedProductId    = try c.decodeIfPresent(String.self,    forKey: .sealedProductId)
        productType        = try c.decodeIfPresent(String.self,    forKey: .productType)
        packsPerBox        = try c.decodeIfPresent(Int.self,       forKey: .packsPerBox)
        cardsPerPack       = try c.decodeIfPresent(Int.self,       forKey: .cardsPerPack)
        totalCards         = try c.decodeIfPresent(Int.self,       forKey: .totalCards)
        msrp               = try c.decodeIfPresent(Double.self,    forKey: .msrp)
        upc                = try c.decodeIfPresent(String.self,    forKey: .upc)
        highlights         = try c.decodeIfPresent([String].self,  forKey: .highlights)
        caseQuantity       = try c.decodeIfPresent(Int.self,       forKey: .caseQuantity)
        ebaySearchQuery    = try c.decodeIfPresent(String.self,    forKey: .ebaySearchQuery)
    }
}

// MARK: - Transferable (drag-and-drop, DESIGN.md §6.6)

/// Inlined into Card.swift rather than a standalone file because
/// Xcode's PBXFileSystemSynchronizedRootGroup intermittently fails
/// to pick up new files (per memory feedback_xcode_synchronized_groups).
/// CodableRepresentation works because Card is already Codable+Sendable.
///
/// contentType is .json (built-in UTType) — that's what
/// CodableRepresentation actually emits via the default JSONEncoder.
/// An earlier custom UTType (exportedAs a bundle-namespaced string)
/// tripped Xcode's "exported type not declared in Info.plist"
/// warning; for in-app-only drag-drop (pool → editor) we don't need
/// or want a custom UTI announcing that BOBA Playbook can receive
/// cards. NOTE: don't put the literal old UTI string anywhere in
/// this file — Xcode's grep-based UTI scan will flag it again.
///
/// `nonisolated` matches the protocol's nonisolated requirement —
/// without it, Card's default MainActor isolation creates the
/// "Conformance crosses into main actor-isolated code" data-race
/// error in Swift 6 mode.
extension Card: Transferable {
    nonisolated static var transferRepresentation: some TransferRepresentation {
        CodableRepresentation(contentType: .json)
    }
}

extension Card {
    /// Destination for the per-card "View on Radish" external-link
    /// button. Prefers the legacy frozen `radishUrl` field (acquired
    /// before 2026-05-23 from Radish's sitemap, treated as static
    /// reference data); falls back to the Radish homepage when the
    /// field is null. This is the ONE permitted use of Radish-derived
    /// data per the email — no probing, no validation, no lookups,
    /// no pricing impact, no Worker call.
    var radishDisplayURL: URL {
        if let raw = radishUrl, let url = URL(string: raw) { return url }
        return URL(string: "https://radishpriceguide.com")!
    }
}

extension Card {
    /// Canonical card identifier matching `scripts/boba_id.py`'s
    /// 5-field v3 formula:
    ///   `cardNumber-(hero or name)-(treatment or "")-(variation or "")-(element or "")`
    /// The 5th field is the card's WEAPON (catalog stores it under the
    /// legacy field name `element` per DECISIONS.md #027). Added
    /// 2026-05-25 to disambiguate FIRE-weapon vs GLOW-weapon variant
    /// siblings that share otherwise-identical (cardNumber, hero,
    /// treatment, variation). Without weapon in the bobaId, the
    /// catalog needed to use distinct cardNumbers as a workaround for
    /// those pairs.
    /// Sealed products use `name` when `hero` is empty. Trailing
    /// dashes are intentional and stable. CLAUDE.md "One ID per
    /// Card" — this is the primary key for the catalog.
    ///
    /// Computed at runtime rather than decoded — the formula is
    /// deterministic so the value matches the `bobaId` stored in
    /// the JSON bundles. `nonisolated` because `id` (Hashable key)
    /// returns this value and must be callable from any actor.
    nonisolated var bobaId: String {
        let identifier = hero.isEmpty ? name : hero
        return "\(cardNumber)-\(identifier)-\(treatment ?? "")-\(variation ?? "")-\(element)"
    }
}
